### PR TITLE
Consensus refactoring. New ReenterAfterCancellation-abstraction.

### PR DESCRIPTION
**Pull request by mhelleborg**

### CODE REVIEW

File: src/Proto.Actor/Context/ContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+using System.Threading.Tasks;
+
+namespace Proto.Context
+{
+    public static class ContextExtensions
+    {
+        /// <summary>
+        /// Calls the callback on token cancellation. If CancellationToken is non-cancellable, this is a noop.
+        /// </summary>
+        public static CancellationTokenRegistration ReenterAfterCancellation(this IContext context, CancellationToken token, Action onCancelled)
```
Comment (mhelleborg) [inline src/Proto.Actor/Context/ContextExtensions.cs]: An alternative to having it as an extension is to add it directly to context. Then we could perhaps add cleanup for contexts with callbacks still waiting. With this model that is on the user.

File: tests/Proto.Cluster.Tests/GossipCoreTests.cs [hunk 2]
```diff
@@ -66,8 +67,11 @@ void SendState(MemberStateDelta memberStateDelta, Member targetMember, InstanceL
             }
 
             var first = environment.Values.First().Gossip;
-            
-            var handle = RegisterConsensusCheck<ClusterTopology, ulong>("topology", tp => tp.TopologyHash, first);
+
+            var checkDefinition = Gossiper.ConsensusCheckBuilder<ulong>.Create("topology", (ClusterTopology tp) => tp.TopologyHash);
```
Comment (rogeralsing) [inline tests/Proto.Cluster.Tests/GossipCoreTests.cs:71]: Nice

**Comment by rogeralsing** [comment]
Looks great!

**Comment by mhelleborg** [comment]
Great, merging :)

**Comment by mhelleborg** [comment]
Wrong button 😄

**Pull request merged**